part4.md

```diff
@@ -0,0 +1,139 @@
+## Part 4 - Even more Kotlin
+
+ * Create package `src/main/kotlin/bindings` and file `Admin.kt`
+
+ * Write basic JS binding to remove `require` and abstract away module import:
+``` kotlin
+@JsModule("firebase-admin")
+@JsNonModule
+external object Admin {
+    fun initializeApp(config: dynamic)
+    fun database(): dynamic
+}
+```
+
+ * In similar way add another binding in `Functions.kt`:
+``` kotlin
+@JsModule("firebase-functions")
+@JsNonModule
+external object Functions {
+    fun config(): dynamic
+    val https: dynamic
+}
+```
+
+ * Also rewrite some dynamic code in main function with newly written bindings:
+``` kotlin
+fun main(args: Array<String>) {
+    Admin.initializeApp(Functions.config().firebase)
+
+    exports.testPush = Functions.https.onRequest { req, res ->
+        Admin.database().ref("/conversions")
+                .push(InputConversions(req.query.eur.toString().toFloat()))
+                .then { res.status(200).send("done") }
+    }
+}
+```
+
+ * Build, deploy and make sure it still works.
+
+ * Copy the rest of my custom bindings (`bindings` folder in this repo) into `src/main/kotlin/bindings` 
+
+ * **Optional:** Examine copied bindings to see how it represents respective JavaScript APIs.
+
+ * Rewrite database table types to Kotlin:
+``` kotlin
+data class Rates(
+    val usd: Float,
+    val czk: Float,
+    val pln: Float,
+    val rub: Float
+)
+
+data class Conversions(
+    val eur: Float,
+    val usd: Float? = null,
+    val czk: Float? = null,
+    val pln: Float? = null,
+    val rub: Float? = null,
+    val timestamp: Long? = null
+)
+```
+
+ * Rewrite `convert` function to Kotlin:
+``` kotlin
+exports.convert = Functions.database.ref("/conversions/{conversionID}").onWrite<Conversions> { event ->
+    val input = event.data.`val`()
+    if (input.timestamp != null) return@onWrite
+    Admin.database().ref("/rates").once<Rates>("value")
+        .then<Conversions> { ratesSnap ->
```

Comment:
it?

Reply:
Way too many nested blocks unfortunately. This is a bit more readable.